internal/storage/raft/backend.go

```diff
@@ -0,0 +1,223 @@
+package raft
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/internal/storage"
+	"github.com/hashicorp/consul/internal/storage/inmem"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+
+	pbstorage "github.com/hashicorp/consul/proto/private/pbstorage"
+)
+
+// NewBackend returns a storage backend that uses Raft for durable persistence
+// and serves reads from an in-memory database. It's suitable for production use.
+//
+// It's not an entirely clean abstraction because rather than owning the Raft
+// subsystem directly, it has to integrate with the existing FSM and related
+// machinery from before generic resources.
+//
+// The given Handle will be used to apply logs and interrogate leadership state.
+// In certain restricted circumstances, Handle may be nil, such as during tests
+// that only exercise snapshot restoration, or when initializing a throwaway FSM
+// during peers.json recovery - but calling any of the data access methods (read
+// or write) will result in a panic.
+//
+// Calling WriteCAS, DeleteCAS, or ReadConsistent on a follower will result in
+// a storage.ErrInconsistent error. It's expected that for these operations, the
+// whole thing will be forwarded to the leader at the RPC level.
+//
+// You must call Run before using the backend.
+func NewBackend(h Handle) (*Backend, error) {
+	s, err := inmem.NewStore()
+	if err != nil {
+		return nil, err
+	}
+	return &Backend{h, s}, nil
+}
+
+// Handle provides glue for interacting with the Raft subsystem via existing
+// machinery on consul.Server.
+type Handle interface {
+	// Apply the given log message.
+	Apply(msg []byte) (any, error)
```

Comment:
Following up from our call earlier:

The original plan was the forward the entire request (at the RPC level), partly because I'd assumed this was necessary from a security/authz point of view. As you pointed out, this isn't the case! We can use the same TLS certificate authorization that we already use for Raft traffic. I think this would be a much cleaner separation of concerns, so will ticket up a timeboxed spike.